File: sdks/js/src/attributes.ts
import type { Attributes } from '@opentelemetry/api';
import type {
  FrameworkAttributes,
  LLMOperationType,
  LLMSpanAttributes,
  VectorDBAttributes,
  WorkflowAttributes,
} from './types';

/**
 * LLM attribute keys following OpenTelemetry semantic conventions
 */
export const LLM_ATTRIBUTES = {
  COMPLETION_TOKENS: 'llm.completion.tokens',
  COST_COMPLETION: 'llm.cost.completion',
  COST_PROMPT: 'llm.cost.prompt',
  COST_TOTAL: 'llm.cost.total',
  DURATION: 'llm.duration_ms',
  ERROR: 'llm.error',
  ERROR_TYPE: 'llm.error.type',
  MAX_TOKENS: 'llm.max_tokens',
  MODEL: 'llm.model',
  OPERATION_TYPE: 'llm.operation.type',
  PROMPT_TOKENS: 'llm.prompt.tokens',
  PROVIDER: 'llm.provider',
  REQUEST_ID: 'llm.request.id',
  STREAM: 'llm.stream',
  TEMPERATURE: 'llm.temperature',
  TOOL_CALLS: 'llm.tool_calls',
  TOOLS: 'llm.tools',
  TOP_P: 'llm.top_p',
  TOTAL_TOKENS: 'llm.total.tokens',
  USAGE_INPUT_TOKENS: 'llm.usage.input_tokens',
  USAGE_OUTPUT_TOKENS: 'llm.usage.output_tokens',
  USAGE_REASON: 'llm.usage.reason',
  USAGE_TOTAL_TOKENS: 'llm.usage.total_tokens',
} as const;

/**
 * Attribute aliases for common use
 */
export const ATTR_LLM_PROVIDER = LLM_ATTRIBUTES.PROVIDER;
export const ATTR_LLM_MODEL = LLM_ATTRIBUTES.MODEL;

/**
 * Vector DB attribute keys
 */
export const VECTOR_DB_ATTRIBUTES = {
  COLLECTION: 'db.collection',
  COUNT: 'vector.count',
  DIMENSION: 'vector.dimension',
  NAME: 'db.name',
  NAMESPACE: 'db.namespace',
  OPERATION: 'db.operation',
  QUERY_FILTER: 'vector.query.filter',
  QUERY_K: 'vector.query.k',
  QUERY_METRIC: 'vector.query.metric',
  SYSTEM: 'db.system',
} as const;

/**
 * Framework attribute keys
 */
export const FRAMEWORK_ATTRIBUTES = {
  AGENT_NAME: 'framework.agent.name',
  AGENT_TYPE: 'framework.agent.type',
  CHAIN_NAME: 'framework.chain.name',
  CHAIN_TYPE: 'framework.chain.type',
  NAME: 'framework.name',
  OPERATION: 'framework.operation',
  TOOL_NAME: 'framework.tool.name',
  TOOL_TYPE: 'framework.tool.type',
  VERSION: 'framework.version',
} as const;

/**
 * Workflow attribute keys
 */
export const WORKFLOW_ATTRIBUTES = {
  METADATA: 'workflow.metadata',
  NAME: 'workflow.name',
  PARENT_ID: 'workflow.parent_id',
  RUN_ID: 'workflow.run_id',
  SESSION_ID: 'workflow.session_id',
  USER_ID: 'workflow.user_id',
  VERSION: 'workflow.version',
} as const;

/**
 * Create LLM span attributes
 */
export function createLLMAttributes(
  provider: string,
  model: string,
  operationType: LLMOperationType,
  additionalAttributes?: Partial<LLMSpanAttributes>,
): LLMSpanAttributes {
  return {
    [LLM_ATTRIBUTES.PROVIDER]: provider,
    [LLM_ATTRIBUTES.MODEL]: model,
    [LLM_ATTRIBUTES.OPERATION_TYPE]: operationType,
    ...additionalAttributes,
  } as LLMSpanAttributes;
}

/**
 * Create vector DB attributes
 */
export function createVectorDBAttributes(
  system: string,
  operation: string,
  additionalAttributes?: Partial<VectorDBAttributes>,
): VectorDBAttributes {
  return {
    [VECTOR_DB_ATTRIBUTES.SYSTEM]: system,
    [VECTOR_DB_ATTRIBUTES.OPERATION]: operation,
    ...additionalAttributes,
  } as VectorDBAttributes;
}

/**
 * Create framework attributes
 */
export function createFrameworkAttributes(
  name: string,
  operation: string,
  additionalAttributes?: Partial<FrameworkAttributes>,
): FrameworkAttributes {
  return {
    [FRAMEWORK_ATTRIBUTES.NAME]: name,
    [FRAMEWORK_ATTRIBUTES.OPERATION]: operation,
    ...additionalAttributes,
  } as FrameworkAttributes;
}

/**
 * Create workflow attributes
 */
export function createWorkflowAttributes(
  name: string,
  runId: string,
  additionalAttributes?: Partial<WorkflowAttributes>,
): WorkflowAttributes {
  return {
    [WORKFLOW_ATTRIBUTES.NAME]: name,
    [WORKFLOW_ATTRIBUTES.RUN_ID]: runId,
    ...additionalAttributes,
  } as WorkflowAttributes;
}

/**
 * Sanitize attributes by removing undefined values
 */
export function sanitizeAttributes(
  attributes: Record<string, unknown>,
): Attributes {
  const sanitized: Attributes = {};

  for (const [key, value] of Object.entries(attributes)) {
    if (value !== undefined && value !== null) {
      if (typeof value === 'object') {
        sanitized[key] = JSON.stringify(value);
      } else if (
        typeof value === 'string' ||
        typeof value === 'number' ||
        typeof value === 'boolean'
      ) {
        sanitized[key] = value;
      } else {
        sanitized[key] = String(value);
      }
    }
  }

  return sanitized;
}

/**
 * Merge multiple attribute objects
 */
export function mergeAttributes(...attributeSets: Attributes[]): Attributes {
  return Object.assign({}, ...attributeSets);
}

File: sdks/rust/src/attributes.rs
//! Attribute definitions and helpers for the Untrace SDK

use opentelemetry::{Key, KeyValue};
use std::collections::HashMap;

/// LLM-specific attributes
pub mod llm {
    use super::*;

    pub const PROVIDER: &str = "llm.provider";
    pub const MODEL: &str = "llm.model";
    pub const OPERATION: &str = "llm.operation";
    pub const PROMPT_TOKENS: &str = "llm.prompt_tokens";
    pub const COMPLETION_TOKENS: &str = "llm.completion_tokens";
    pub const TOTAL_TOKENS: &str = "llm.total_tokens";
    pub const TEMPERATURE: &str = "llm.temperature";
    pub const TOP_P: &str = "llm.top_p";
    pub const MAX_TOKENS: &str = "llm.max_tokens";
    pub const STREAM: &str = "llm.stream";
    pub const TOOLS: &str = "llm.tools";
    pub const TOOL_CALLS: &str = "llm.tool_calls";
    pub const DURATION_MS: &str = "llm.duration_ms";
    pub const COST_PROMPT: &str = "llm.cost_prompt";
    pub const COST_COMPLETION: &str = "llm.cost_completion";
    pub const COST_TOTAL: &str = "llm.cost_total";
    pub const ERROR: &str = "llm.error";
    pub const ERROR_TYPE: &str = "llm.error_type";
    pub const REQUEST_ID: &str = "llm.request_id";
    pub const USAGE_REASON: &str = "llm.usage_reason";
}

/// Vector database attributes
pub mod vector_db {
    use super::*;

    pub const PROVIDER: &str = "vector_db.provider";
    pub const COLLECTION: &str = "vector_db.collection";
    pub const OPERATION: &str = "vector_db.operation";
    pub const DIMENSIONS: &str = "vector_db.dimensions";
    pub const VECTOR_COUNT: &str = "vector_db.vector_count";
    pub const QUERY_VECTOR_COUNT: &str = "vector_db.query_vector_count";
    pub const RESULT_COUNT: &str = "vector_db.result_count";
    pub const SIMILARITY_THRESHOLD: &str = "vector_db.similarity_threshold";
    pub const FILTER: &str = "vector_db.filter";
    pub const METADATA: &str = "vector_db.metadata";
}

/// Framework attributes
pub mod framework {
    use super::*;

    pub const NAME: &str = "framework.name";
    pub const VERSION: &str = "framework.version";
    pub const TYPE: &str = "framework.type";
    pub const OPERATION: &str = "framework.operation";
    pub const COMPONENT: &str = "framework.component";
    pub const METHOD: &str = "framework.method";
    pub const ROUTE: &str = "framework.route";
    pub const STATUS_CODE: &str = "framework.status_code";
    pub const DURATION_MS: &str = "framework.duration_ms";
    pub const ERROR: &str = "framework.error";
    pub const ERROR_TYPE: &str = "framework.error_type";
}

/// Workflow attributes
pub mod workflow {
    use super::*;

    pub const ID: &str = "workflow.id";
    pub const NAME: &str = "workflow.name";
    pub const RUN_ID: &str = "workflow.run_id";
    pub const USER_ID: &str = "workflow.user_id";
    pub const SESSION_ID: &str = "workflow.session_id";
    pub const VERSION: &str = "workflow.version";
    pub const PARENT_ID: &str = "workflow.parent_id";
    pub const STATUS: &str = "workflow.status";
    pub const DURATION_MS: &str = "workflow.duration_ms";
    pub const ERROR: &str = "workflow.error";
    pub const ERROR_TYPE: &str = "workflow.error_type";
    pub const METADATA: &str = "workflow.metadata";
}

/// Create LLM attributes from a map
pub fn create_llm_attributes(attrs: &HashMap<String, String>) -> Vec<KeyValue> {
    attrs
        .iter()
        .map(|(k, v)| KeyValue::new(Key::new(k.clone()), v.clone()))
        .collect()
}

/// Create vector database attributes from a map
pub fn create_vector_db_attributes(attrs: &HashMap<String, String>) -> Vec<KeyValue> {
    attrs
        .iter()
        .map(|(k, v)| KeyValue::new(Key::new(k.clone()), v.clone()))
        .collect()
}

/// Create framework attributes from a map
pub fn create_framework_attributes(attrs: &HashMap<String, String>) -> Vec<KeyValue> {
    attrs
        .iter()
        .map(|(k, v)| KeyValue::new(Key::new(k.clone()), v.clone()))
        .collect()
}

/// Create workflow attributes from a map
pub fn create_workflow_attributes(attrs: &HashMap<String, String>) -> Vec<KeyValue> {
    attrs
        .iter()
        .map(|(k, v)| KeyValue::new(Key::new(k.clone()), v.clone()))
        .collect()
}

/// Sanitize attributes by removing sensitive information
pub fn sanitize_attributes(attrs: &mut HashMap<String, String>) {
    let sensitive_keys = [
        "password",
        "secret",
        "token",
        "key",
        "auth",
        "credential",
        "api_key",
        "access_token",
        "refresh_token",
    ];

    for key in sensitive_keys {
        if attrs.contains_key(key) {
            attrs.insert(key.to_string(), "[REDACTED]".to_string());
        }
    }
}

/// Merge two attribute maps
pub fn merge_attributes(
    mut base: HashMap<String, String>,
    other: HashMap<String, String>,
) -> HashMap<String, String> {
    for (k, v) in other {
        base.insert(k, v);
    }
    base
}

/// Helper functions for creating common attribute values
pub mod helpers {
    use super::*;

    /// Create a string attribute
    pub fn string(key: &str, value: &str) -> KeyValue {
        KeyValue::new(Key::new(key.to_string()), value.to_string())
    }

    /// Create an integer attribute
    pub fn int(key: &str, value: i64) -> KeyValue {
        KeyValue::new(Key::new(key.to_string()), value)
    }

    /// Create a float attribute
    pub fn float(key: &str, value: f64) -> KeyValue {
        KeyValue::new(Key::new(key.to_string()), value)
    }

    /// Create a boolean attribute
    pub fn bool(key: &str, value: bool) -> KeyValue {
        KeyValue::new(Key::new(key.to_string()), value)
    }

    /// Create a string slice attribute
    pub fn string_slice(key: &str, value: Vec<String>) -> KeyValue {
        KeyValue::new(Key::new(key.to_string()), value.join(","))
    }

    /// Create an integer slice attribute
    pub fn int_slice(key: &str, value: Vec<i64>) -> KeyValue {
        KeyValue::new(Key::new(key.to_string()), value.iter().map(|v| v.to_string()).collect::<Vec<_>>().join(","))
    }

    /// Create a float slice attribute
    pub fn float_slice(key: &str, value: Vec<f64>) -> KeyValue {
        KeyValue::new(Key::new(key.to_string()), value.iter().map(|v| v.to_string()).collect::<Vec<_>>().join(","))
    }
}
